### PR TITLE
rgw_sal_motr: [CORTX-32487]  update stats behavior on multipart upload

### PR DESCRIPTION
**Pull request by sanalpk**
Update Stats behaviour for multipart upload sessions

#### Earlier Behaviour
Increment object count in a bucket per part upload in a multipart session.
At the multipart complete stage adjust the object count so that only one
object is added in the end.

#### Problem
If the user quota is of 3 objects then the user can not upload more than 3 parts 
for a single multipart object. (This is because we are updating the object count
for each part.)

#### Resolution
Increment the object count only at a multipart init stage. 
Increment the object size on the upload of each part. (no update for obj_count).

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -3756,7 +3754,7 @@ int MotrMultipartUpload::delete_parts(const DoutPrefixProvider *dpp, std::string
     // Subtract size & count of all the parts if multipart is not completed.
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:3754]: Please update the comment here.

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -3922,10 +3920,18 @@ int MotrMultipartUpload::init(const DoutPrefixProvider *dpp, optional_yield y,
...
   if (rc == -EEXIST)
     rc = 0;
-  if (rc < 0)
+  if (rc < 0) {
     // TODO: clean the bucket index entry
     ldpp_dout(dpp, 0) << "Failed to create object multipart index  " << obj_part_iname << dendl;
+    return rc;
+  }
 
+  rc = update_bucket_stats(dpp, store, owner.get_id().to_str(), tenant_bkt_name, 0, 0, 1, true);
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:3929]: add comment above

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -4268,19 +4274,6 @@ int MotrMultipartUpload::complete(const DoutPrefixProvider *dpp,
   store->get_obj_meta_cache()->put(dpp, tobj_key, update_bl);
 
   // Increment size & count for new multipart obj in bucket stats entry.
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:4276]: remove the comment

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -4498,8 +4489,7 @@ int MotrMultipartWriter::complete(size_t accounted_size, const std::string& etag
                            head_obj->get_bucket()->get_owner()->get_id().to_str(),
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:4489]: pls, do add comments above this update call as well.

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (sanalpk) [reply src/rgw/rgw_sal_motr.cc:3930]: done

File: src/rgw/rgw_sal_motr.cc (hunk 4 above)
Comment (sanalpk) [reply src/rgw/rgw_sal_motr.cc:4490]: done

**Review by sachinpunadikar — APPROVED** [review]
LGTM

File: src/rgw/rgw_sal_motr.cc [hunk 5]
```diff
@@ -4493,13 +4484,13 @@ int MotrMultipartWriter::complete(size_t accounted_size, const std::string& etag
     ldpp_dout(dpp, 0) <<__func__<< ": failed to add part obj in part index, rc=" << rc << dendl;
     return rc == -ENOENT ? -ERR_NO_SUCH_UPLOAD : rc;
   }
-
-   rc = update_bucket_stats(dpp, store,
+   
+  //update size without changing the object count
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:4488]:
```suggestion
  // update size without changing the object count
```

**Review by jjxsg — APPROVED** [review]
@sanalpk Please resolve conflicts and update the tests performed.

**Comment by sumedhak27** [comment]
@sanalpk, I have updated the PR description. Please update it in the commit-msg.

**Comment by sanalpk** [comment]
Test
------
Before multi part upload

Request:GET http://ssc-vm-g3-rhev4-2883.colo.seagate.com:8000/admin/bucket?stats&uid=statuser002&format=json HTTP/1.1
HTTP/1.1 200 OK

Response Content is: [{"bucket":"bucketstatuser","num_shards":1,"tenant":"","zonegroup":"0956b174-fe14-4f97-8b50-bb7ec5e1cf62","placement_rule":"default","explicit_placement":

{"data_pool":"","data_extra_pool":"","index_pool":""}
,"id":"","marker":"","index_type":"Normal","owner":"statuser002","ver":"","master_ver":"","mtime":"0.000000","creation_time":"0.000000","max_marker":"","usage":{"rgw.main":{"size":0,"size_actual":0,"size_utilized":0,"size_kb":0,"size_kb_actual":0,"size_kb_utilized":0,"num_objects":1}},"bucket_quota":{"enabled":false,"check_on_raw":false,"max_size":-1,"max_size_kb":0,"max_objects":-1}},{"bucket":"bucketstatuser2","num_shards":1,"tenant":"","zonegroup":"0956b174-fe14-4f97-8b50-bb7ec5e1cf62","placement_rule":"default","explicit_placement":

{"data_pool":"","data_extra_pool":"","index_pool":""}
,"id":"","marker":"","index_type":"Normal","owner":"statuser002","ver":"","master_ver":"","mtime":"0.000000","creation_time":"0.000000","max_marker":"","usage":{},"bucket_quota":{"enabled":false,"check_on_raw":false,"max_size":-1,"max_size_kb":0,"max_objects":-1}}]

after multi part
----------------
Request:GET http://ssc-vm-g3-rhev4-2883.colo.seagate.com:8000/admin/bucket?stats&uid=statuser002&format=json HTTP/1.1
HTTP/1.1 200 OK

Response Content is: [{"bucket":"bucketstatuser","num_shards":1,"tenant":"","zonegroup":"0956b174-fe14-4f97-8b50-bb7ec5e1cf62","placement_rule":"default","explicit_placement":

{"data_pool":"","data_extra_pool":"","index_pool":""}
,"id":"","marker":"","index_type":"Normal","owner":"statuser002","ver":"","master_ver":"","mtime":"0.000000","creation_time":"0.000000","max_marker":"","usage":{"rgw.main":{"size":872415232,"size_actual":872415232,"size_utilized":0,"size_kb":851968,"size_kb_actual":851968,"size_kb_utilized":0,"num_objects":2}},"bucket_quota":{"enabled":false,"check_on_raw":false,"max_size":-1,"max_size_kb":0,"max_objects":-1}},{"bucket":"bucketstatuser2","num_shards":1,"tenant":"","zonegroup":"0956b174-fe14-4f97-8b50-bb7ec5e1cf62","placement_rule":"default","explicit_placement":

{"data_pool":"","data_extra_pool":"","index_pool":""}
,"id":"","marker":"","index_type":"Normal","owner":"statuser002","ver":"","master_ver":"","mtime":"0.000000","creation_time":"0.000000","max_marker":"","usage":{},"bucket_quota":{"enabled":false,"check_on_raw":false,"max_size":-1,"max_size_kb":0,"max_objects":-1}}]

File: src/rgw/rgw_sal_motr.cc (hunk 5 above)
Comment (sanalpk) [reply src/rgw/rgw_sal_motr.cc]: done

**Review by vastradparayya — APPROVED** [review]
Please resolve conflict

File: src/rgw/rgw_sal_motr.cc [hunk 6]
```diff
@@ -3926,10 +3924,19 @@ int MotrMultipartUpload::init(const DoutPrefixProvider *dpp, optional_yield y,
...
   if (rc == -EEXIST)
     rc = 0;
-  if (rc < 0)
+  if (rc < 0) {
     // TODO: clean the bucket index entry
     ldpp_dout(dpp, 0) << "Failed to create object multipart index  " << obj_part_iname << dendl;
+    return rc;
+  }
 
+  // initialize with size=0 & object count=1
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:3933]:
```suggestion
  // Add "1" to the object_count of the current bucket stats.
  // Size will be added when parts are uploaded.
```

File: src/rgw/rgw_sal_motr.cc [hunk 7]
```diff
@@ -3926,10 +3924,19 @@ int MotrMultipartUpload::init(const DoutPrefixProvider *dpp, optional_yield y,
...
+  if (rc < 0) {
     // TODO: clean the bucket index entry
     ldpp_dout(dpp, 0) << "Failed to create object multipart index  " << obj_part_iname << dendl;
+    return rc;
+  }
 
+  // initialize with size=0 & object count=1 
+  rc = update_bucket_stats(dpp, store, owner.get_id().to_str(), tenant_bkt_name, 0, 0, 1, true);
+  if (rc != 0) {
+    ldpp_dout(dpp, 20) <<__func__<< ": Failed stats substraction for the "
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:3936]: please update the comment msg. this is not stats subtraction.
Msg can be `Failed to update object count for the bucket/obj=**`

**Review by jjxsg — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 6 above)
Comment (sanalpk) [reply src/rgw/rgw_sal_motr.cc]: done

File: src/rgw/rgw_sal_motr.cc (hunk 7 above)
Comment (sanalpk) [reply src/rgw/rgw_sal_motr.cc]: done

**Review by sumedhak27 — APPROVED** [review]

**Pull request merged**